Relieve db pressure on api/health check

dpsutton opened the pull request:
https://github.com/metabase/metabase/issues/26266

Servers under heavy load can be slow to respond to the api/health check. This can lead to k8s killing healthy instances happily humming along serving requests.

One idea floated was to use QoSFilters
https://www.eclipse.org/jetty/javadoc/jetty-9/org/eclipse/jetty/servlets/QoSFilter.html to prioritize those requests in front of others. But I suspect this might not be our bottleneck.

Our health endpoint was updated to see if it could acquire an endpoint when we were dealing with connection pool issues. We were reporting the instance was healthy once it has finished the init process, but would report healthy if 60/15 app-db connections were used and no actual queries could complete.

The remedy was adding
`(sql-jdbc.conn/can-connect-with-spec? {:datasource (mdb.connection/data-source)})` to the endpoint. But now to get information about the health of the system we have to wait in the queue to get a db connection.

The hope is that this change which monitors for recent db checkins (query success) and checkouts (query begun) can be a proxy for db activity without having to wait for a connection and hit the db ourselves.

Some simple and crude benchmarking:
- use `siege` to hit `api/database/<app-db>/sync_schema`
- in a separate tab, use `siege` to hit `api/health`

Three trials with unconditional db access and conditional db access (look for recent activity set by the new `ConnectionCustomizer`).

One siege client is synching the app-db's schema with 80 clients each sending 60 requests. the other has 1 client sending 60 requests to api/health.

```
Run             |  Elapsed Time | max tx  | tx rate
 before change  |    7.16s      |  0.79s  |  8.38 tx/s
 before change  |   23.91s      |  1.44s  |  2.51 tx/s
 before change  |   13.00s      |  0.50s  |  4.62 tx/s
----------------------------------------------------
 after change   |    4.46s      |  0.27s  |  13.45 tx/s
 after change   |    5.81s      |  0.61s  |  10.33 tx/s
 after change   |    4.54s      |  0.44s  |  13.22 tx/s
```

Full(er) results below:

```
Unconditional db access
=======================

siege -c80 -r 40 "http://localhost:3000/api/database/2/sync_schema POST" -H "Cookie: $SESSION"

siege -c 1 -r 60 "http://localhost:3000/api/health"

Elapsed time:		        7.16 secs
Response time:		        0.12 secs
Transaction rate:	        8.38 trans/sec
Longest transaction:	        0.79
Shortest transaction:	        0.01

Elapsed time:		       23.91 secs
Response time:		        0.40 secs
Transaction rate:	        2.51 trans/sec
Longest transaction:	        1.44
Shortest transaction:	        0.02

Elapsed time:		       13.00 secs
Response time:		        0.22 secs
Transaction rate:	        4.62 trans/sec
Longest transaction:	        0.50
Shortest transaction:	        0.06

Conditional db access
==============================================================

Elapsed time:		        4.46 secs
Response time:		        0.07 secs
Transaction rate:	       13.45 trans/sec
Longest transaction:	        0.27
Shortest transaction:	        0.01

Elapsed time:		        5.81 secs
Response time:		        0.10 secs
Transaction rate:	       10.33 trans/sec
Longest transaction:	        0.61
Shortest transaction:	        0.00

Elapsed time:		        4.54 secs
Response time:		        0.08 secs
Transaction rate:	       13.22 trans/sec
Longest transaction:	        0.44
Shortest transaction:	        0.01
```


I'm not wild that we have to use reflection to add our connection customizer to c3p0. I've got an issue open at https://github.com/swaldman/c3p0/issues/166 to see if that can be made better.

I tried several ways to get our class on c3p0's classpath.
- `(classloader/the-classloader)` doesn't work because the class is looked up from a thread launched by c3p0 which has the application classloader and not clojure's dynamic one. We are not in control of the creation of this thread so i'm not sure how we could swap out its classloader
- we could make a java class and compile it and put it on our classpath. presumably c3p0 would see this. but this would bring us back to the world of having a prep step and we don't want to deal with that if we don't have to
- trying to figure out other classloader shenanigans but ultimately to no avail.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/metabase/metabase/27192)
<!-- Reviewable:end -->

dpsutton committed:
Relieve db pressure on api/health check

https://github.com/metabase/metabase/issues/26266

Servers under heavy load can be slow to respond to the api/health
check. This can lead to k8s killing healthy instances happily humming
along serving requests.

One idea floated was to use QoSFilters
https://www.eclipse.org/jetty/javadoc/jetty-9/org/eclipse/jetty/servlets/QoSFilter.html
to prioritize those requests in front of others. But I suspect this
might not be our bottleneck.

Our health endpoint was updated to see if it could acquire an endpoint
when we were dealing with connection pool issues. We were reporting the
instance was healthy once it has finished the init process, but would
report healthy if 60/15 app-db connections were used and no actual
queries could complete.

The remedy was adding
`(sql-jdbc.conn/can-connect-with-spec? {:datasource (mdb.connection/data-source)})`
to the endpoint. But now to get information about the health of the
system we have to wait in the queue to get a datasource.

The hope is that this change which monitors for recent db
checkins (query success) and checkouts (query begun) can be a proxy for
db activity without having to wait for a connection and hit the db ourselves.

Some simple and crude benchmarking:
- use `siege` to hit `api/database/<app-db>/sync_schema`
- in a separate tab, use `siege` to hit `api/health`

Three trials with unconditional db access and conditional db
access (look for recent activity set by the new `ConnectionCustomizer`).

One siege client is synching the app-db's schema with 80 clients each
sending 60 requests. the other has 1 client sending 60 requests to api/health.

Run             |  Elapsed Time | max tx  | tx rate
 before change  |    7.16s      |  0.79s  |  8.38 tx/s
 before change  |   23.91s      |  1.44s  |  2.51 tx/s
 before change  |   13.00s      |  0.50s  |  4.62 tx/s
----------------------------------------------------
 after change   |    4.46s      |  0.27s  |  13.45 tx/s
 after change   |    5.81s      |  0.61s  |  10.33 tx/s
 after change   |    4.54s      |  0.44s  |  13.22 tx/s

Full(er) results below:

```
Unconditional db access
=======================

siege -c80 -r 40 "http://localhost:3000/api/database/2/sync_schema POST" -H "Cookie: $SESSION"

siege -c 1 -r 60 "http://localhost:3000/api/health"

Elapsed time:		        7.16 secs
Response time:		        0.12 secs
Transaction rate:	        8.38 trans/sec
Longest transaction:	        0.79
Shortest transaction:	        0.01

Elapsed time:		       23.91 secs
Response time:		        0.40 secs
Transaction rate:	        2.51 trans/sec
Longest transaction:	        1.44
Shortest transaction:	        0.02

Elapsed time:		       13.00 secs
Response time:		        0.22 secs
Transaction rate:	        4.62 trans/sec
Longest transaction:	        0.50
Shortest transaction:	        0.06

Conditional db access
==============================================================

Elapsed time:		        4.46 secs
Response time:		        0.07 secs
Transaction rate:	       13.45 trans/sec
Longest transaction:	        0.27
Shortest transaction:	        0.01

Elapsed time:		        5.81 secs
Response time:		        0.10 secs
Transaction rate:	       10.33 trans/sec
Longest transaction:	        0.61
Shortest transaction:	        0.00

Elapsed time:		        4.54 secs
Response time:		        0.08 secs
Transaction rate:	       13.22 trans/sec
Longest transaction:	        0.44
Shortest transaction:	        0.01
```

dpsutton committed:
Remove reflection in `.put` call (not the reflections trategy)

also remove the call to `classloader/the-classloader` as it did nothing

camsaul commented on src/metabase/db/connection_pool_setup.clj:
```diff
@@ -1,17 +1,45 @@
+  (onCheckIn [_ _connection _identity-token]
+    (reset! latest-checkin (t/offset-date-time)))
+  (onCheckOut [_ _connection _identity-token]
+    (reset! latest-checkin (t/offset-date-time)))
```
Was this left in unintentionally? Or maybe the docstring for `recent-activity?` and the atom name need tweaking?

dpsutton replied:
Which strikes you as best? I'm leaning towards making everything keyed off of check in but open to other ideas.

camsaul commented on src/metabase/db/connection_pool_setup.clj:
```diff
@@ -1,17 +1,45 @@
+    (reset! latest-checkin (t/offset-date-time)))
+  (onDestroy [_ _connection _identity-token]))
+
+(let [field (doto (.getDeclaredField com.mchange.v2.c3p0.C3P0Registry "classNamesToConnectionCustomizers")
```
This sort of shenanigan at the very least deserves a comment explaining why you're doing it. A link to your upstream c3p0 issue would be nice too

dpsutton replied:
good call thanks

camsaul commented on src/metabase/db/connection_pool_setup.clj:
```diff
@@ -1,17 +1,45 @@
+
+(def ^:private latest-checkin (atom nil))
+
+(def ^:private recent-window-seconds 15)
```
```suggestion
(def ^:private ^java.time.Duration recent-window (t/seconds 15))
```

this will facilitate testing this (more on this in a second)

camsaul commented on src/metabase/db/connection_pool_setup.clj:
```diff
@@ -1,17 +1,45 @@
+
+(def ^:private recent-window-seconds 15)
+
+(defn recent-activity?
+  "Returns true if there has been recent activity. Define recent activity as an application db connection checked in
+  within 15 seconds. Check-in means a query succeeded and the db connection is no longer needed."
+  []
+  (when-let [activity @latest-checkin]
+    (t/after? activity (t/minus (t/offset-date-time) (t/seconds recent-window-seconds)))))
```
We should write a test for this

- `with-redefs` the `recent-window` to something super short like 1 millisecond
   - Thread sleep for 2 milliseconds just to be extra safe
   - `recent-activity?` should return `false`
- (outside of the `with-redefs`)
   - check out a connection from the app DB (`jdbc/with-db-connection [_conn (db/connection)]` or something would work 
   - `recent-activity?` should return `true` now

camsaul reviewed (approved):
LGTM overall but 

- you might want to clarify the dox that `recent-activity?` records either check in **or** check out.
- shady reflective access deserves an explanatory comment
- a test would be nice

dpsutton committed:
Comment and settle on a single method

dpsutton committed:
tests

dpsutton committed:
select from db twice

had a failure in CI. give it time to do its thing with another db call

dpsutton committed:
block to wait for timestamp update?

dpsutton committed:
unflake the tests

tasks and events from outside the thread can hit the db. the
ConnectionCustomizer is also run from c3p0 controlled threads so we
can't easily isolate everything to our thread

Was running

```clojure
(comment
  (dotimes [n 5]
    (dotimes [_ 100]
      (recent-activity-test)
      (CheckinTracker-test))
    (println (* (inc n) 100)))
    )
```

to run the tests 500 times and would keep getting flakes at a rate
~1/100 to 1/500. Just frustration for the future.

dpsutton committed:
typehint

dpsutton committed:
Switch it up a bit

Tests were flaking in h2 and I don't know why. I'm switching to just
updating recent activity on most methods.
